libc/src/__support/CPP/array.h

```diff
@@ -45,9 +48,29 @@ template <class T, size_t N> struct array {
 
   LIBC_INLINE constexpr iterator begin() { return Data; }
   LIBC_INLINE constexpr const_iterator begin() const { return Data; }
+  LIBC_INLINE constexpr const_iterator cbegin() const { return begin(); }
 
   LIBC_INLINE constexpr iterator end() { return Data + N; }
   LIBC_INLINE constexpr const_iterator end() const { return Data + N; }
+  LIBC_INLINE constexpr const_iterator cend() const { return end(); }
+
+  LIBC_INLINE constexpr reverse_iterator rbegin() {
+    return reverse_iterator{end()};
+  }
+  LIBC_INLINE constexpr const_reverse_iterator rbegin() const {
+    return const_reverse_iterator{end()};
+  }
+  LIBC_INLINE constexpr const_reverse_iterator crbegin() const {
+    return rbegin();
+  }
+
+  LIBC_INLINE constexpr reverse_iterator rend() {
+    return reverse_iterator{begin()};
+  }
+  LIBC_INLINE constexpr const_reverse_iterator rend() const {
+    return const_reverse_iterator{begin()};
+  }
+  LIBC_INLINE constexpr const_reverse_iterator crend() const { return rend(); }
```

Comment:
Most functions here should be `[[nodiscard]]` I believe.

Reply:
https://en.cppreference.com/w/cpp/container/array/rend also doesn't mention nodiscard.

I guess I'm happy to add it, regardless, since nodiscard should have been the language default IMO.

Reply:
C++17 S23.3.7.1 Class template array overview also doesn't mention [[nodiscard]] on these methods.

Reply:
Hmm...now I'm on the fence.  I'm going to land these without, but I would not be opposed to adding them in a future PR.

Reply:
Coincidentally, I think some WG21 are in Tokyo right now, and this came up for discussion. @lntue pointed me to a discussion in #libcxx channel of llvm discord where Mark de Wever said:

> Going forward it seems LEWG is happy to leave nodiscard to  implementors 🙂  @philnik regarding your nodiscard ideas, did you ask MSVC STL what they do? @ldionne can you put nodiscard on the agenda for our next meeting?

Reply:
Hehe! I didn't mean to confuse you with `[[nodiscard]]` but I'm amazed by the outcome!

Tangentially I think we're missing the following functions for `reverse_iterator` to be useful.
```
  LIBC_INLINE friend constexpr bool operator==(const reverse_iterator &lhs,
                                               const reverse_iterator &rhs) {
    return lhs.current == rhs.current;
  }
  LIBC_INLINE friend constexpr bool operator!=(const reverse_iterator &lhs,
                                               const reverse_iterator &rhs) {
    return lhs.current != rhs.current;
  }
```

as in `for (auto itr = value.rbegin(); itr **!=** value.rend(); ++itr)`.